### ISSUE TITLE
Could you replace "axios" with fetch?

### ISSUE THREAD

**Issue by jimmywarting** [twilio/twilio-node#898]
... as part of #728 ?

let users bring in there own fetch implementation.

**Comment by beebzz** [comment]
Hey @jimmywarting, we'll add this to our internal backlog to be prioritized. +1s on this issue and PRs will bump it higher up the backlog.

**Comment by mariomui** [comment]
@jimmywarting I don't see the point of getting rid of axios.
Axios client has other stuff like httpsAgent and a dependency of Request which has a large surface. I guess you could map a adapter interface so you could use undici but...there are prolly bigger fish to fry. 

It's also super small as a dependency. 
https://github.com/mariomui/twilio-node/tree/smaller-build.   <---I shaved off 100kb to make it 1.8mb but someone else is gonna need to test this a lambda project handy.

In my opinion:

* Deprecate the pattern of : `exports = initializer` 
   * Basically amounts to -> `module.exports = initialzer` ---> Which can be done by babel.
   * also -> we can get rid of the plugin  "babel-plugin-replace-ts-export-assignment"  
* use tsc to generate the types only, export them into dist/interfaces/ 
* use babel to transpile typescript down to js. (performance)
* convert existing code to get to ESM status.
* (wait for someone higher up to spend money to rearchitect 4.0.0 to a monorepos with multiple packages)

**Comment by bnb** [comment]
At the end of April, Node.js v14 will be EOL and v16 will be the minimum supported version of Node.js. [unidici-fetch](https://github.com/nodejs/undici#undicifetchinput-init-promise) fully supports Node.js v16 (I believe some features are missing from v14 that undici-fetch uses). Undici is an official project _under_ the Node.js project, so it will be very will supported within Node.js, and problems will get resolved rapidly as they come up. Undici also generally has a pretty hardcore focus on performance, which will very likely benefit us here.

undici-fetch is directly included in Node.js v18 as Node.js's official fetch implementation, meaning that _next year_ a third-party dependency could be _entirely_ dropped since Node.js will support fetch directly, natively, on all supported release lines.

Outside of adopting undici-fetch, I do +1 allowing users to bring their own fetch implementation if they don't want to use the one we provide. There is already support for a [Custom HTTP Client](https://www.twilio.com/docs/libraries/node/custom-http-clients-node) (thank you @beebzz for pointing this out!), so I'm not sure how different "custom fetch" (request in the OP) and "custom HTTP client" are in practice.

**Comment by aaronhuggins-carewell** [comment]
Absolutely would love to see `fetch` used instead of axios.

**Comment by nileshtrivedi** [comment]
+1 IF this makes twilio usable in Deno environments instead of NodeJS.

**Comment by jimmywarting** [comment]
Same could be said for Bun.js

**Comment by nileshtrivedi** [comment]
There's currently [just one Twilio-related package listed on Deno.land](https://deno.land/x/twilio@0.1.1) which [hasn't seen any activity for almost 2 years now](https://github.com/Dygear/Twilio).

**Comment by UrielCh** [comment]
> There's currently [just one Twilio-related package listed on Deno.land](https://deno.land/x/twilio@0.1.1) which [hasn't seen any activity for almost 2 years now](https://github.com/Dygear/Twilio).

It won't do much with a single file having less than 50 lines.

by the way, this project is only exported as CJS.

**Comment by sarankumar-ns** [comment]
+1

**Comment by paperless** [comment]
Also had issues with this. Using bun and the promise was never resolving for messages.create

**Comment by riley** [comment]
+1 Unless I'm missing something, twilio-node doesn't work in my Electron app at all (node.js context). I would like to use twilio as a platform, but this is obviously a dealbreaker.

twilio gives this error when trying to run the Hello World code after signing up.
```
client.calls.create({
      url: "http://demo.twilio.com/docs/voice.xml",
      to: "<my_number>",
      from: "<my_twilio_number>",
})
 .then(call => console.log(call.sid));

UnhandledPromiseRejectionWarning: AxiosError: There is no suitable adapter to dispatch the request since :
- adapter xhr is not supported by the environment
- adapter http is not available in the build
```

**Comment by OrRosenblatt** [comment]
+1

**Comment by ibash** [comment]
except for typescript nonsense, this seems to work:

set a custom request client but with twilio's type
```typescript
  import twilio, { type RequestClient as TwilioRequestClient } from 'twilio'

const  client = twilio(
    process.env.TWILIO_ACCOUNT_SID!,
    process.env.TWILIO_AUTH_TOKEN!,
    {
      httpClient: new RequestClient() as TwilioRequestClient
    }
  )
```

And a custom request client implementation:
```typescript
// axios (which twilio uses) breaks in bun, so override with fetch by having our
// own RequestClient implementation

// couldn't figure out out how to import the types of RequestOptions and Response
// from the twilio sdk, so copied it
type HttpMethod = 'get' | 'post' | 'put' | 'patch' | 'delete'
interface Headers {
  [header: string]: string
}

interface RequestOptions<TData = any, TParams = object> {
  method: HttpMethod
  uri: string
  username?: string
  password?: string
  headers?: Headers
  params?: TParams
  data?: TData
  timeout?: number
  allowRedirects?: boolean
  forever?: boolean
  logLevel?: string
}

class Response<TPayload> {
  constructor(
    public statusCode: number,
    public body: TPayload,
    public headers: any
  ) {}

  toString(): string {
    return 'HTTP ' + this.statusCode + ' ' + this.body
  }
}

export default class RequestClient {

  async request<TData>(opts: RequestOptions<TData>): Promise<Response<TData>> {
    console.log('got a request', opts)
    const { uri, method, username, password, data, ...rest } = opts

    let headers = opts.headers || {}

    // from twilio sdk
    if (opts.username && opts.password) {
      const auth = Buffer.from(opts.username + ':' + opts.password).toString(
        'base64'
      )
      headers.Authorization = 'Basic ' + auth
    }

    let body: URLSearchParams | string | null = null
    if (
      headers['Content-Type'] === 'application/x-www-form-urlencoded' &&
      data
    ) {
      // TODO(ibash) typescript nonsense
      // Type 'import("url").URLSearchParams' is not assignable to type 'URLSearchParams'.
      body = new URLSearchParams(data) as any
    } else if (headers['Content-Type'] === 'application/json' && data) {
      body = JSON.stringify(data)
    }

    const response = await fetch(uri, {
      // TODO(ibash) uppercase method?
      method,
      headers: headers,
      body: body
    })

    // TODO(ibash) handle other response types
    let responseBody: TData | null = null
    const contentType = response.headers.get('Content-Type')
    if (contentType?.includes('application/json')) {
      responseBody = (await response.json()) as TData
    } else {
      throw new Error(`Unhandled response type: ${contentType}`)
    }

    return new Response(
      response.status,
      responseBody as TData,
      response.headers
    )
  }
}
```

**Comment by mfulton26** [comment]
[`ky`](https://github.com/sindresorhus/ky) can also be used to make working with `fetch()` easier and handling common patterns (hooks, errors, etc.)

**Comment by tawanaj** [comment]
+1

**Comment by rvion** [comment]
yes please ! https://news.ycombinator.com/item?id=47582220 
axios is really legacy at that point ; please 🙏
